File: PrototipoAkinator.m
clear; close all; clc;
%Cargado de base de datos
PPelo = "¿Su color de Pelo es ";
PDeporte = "¿Su Deporte favorito es ";
PSexo = "¿Su sexo es ";
PInterior = "¿Vive en el Interior?";

Opciones = {"Rubio", "Marron", "Morocho"; ...
            "Futbol", "Rugby", "Tennis"; ...
            "Masculino", "Femenino", "Invalid"; ...
            "Si", "No", "Invalid"};

Preguntas = {PPelo, PDeporte, PSexo,PInterior};

Tabla = ['Nombre', 'Pelo', 'Deporte', 'Sexo', 'Interior'; ...
         'Juan', Opciones(1,1), Opciones(2,1), Opciones(3,1), Opciones(4,2); ...
         'Pablo', Opciones(1,2), Opciones(2,2), Opciones(3,1), Opciones(4,1); ...
         'Maria', Opciones(1,3), Opciones(2,3), Opciones(3,2), Opciones(4,2); ...
         'Guillermina', Opciones(1,1), Opciones(2,3), Opciones(3,2), Opciones(4,1); ...
         'Ricardo', Opciones(1,2), Opciones(2,1), Opciones(3,1), Opciones(4,1)];

Repeticiones = zeros(size(Opciones));
for i=1:size(Opciones,1)
  for j=1:size(Opciones,2)
    a = strcmp(Opciones(i,j),Tabla);
    Repeticiones(i,j) = sum(sum(a));
  endfor
endfor

estado = Repeticiones;
candidatos = Tabla(2:6, 1:5);

while size(candidatos,1) > 1
  n = size(candidatos,1);
  posCaract = elegirCaracteristica(estado, n);
  caracteristica = Opciones(posCaract(1), posCaract(2));
  
  if(posCaract(2)==4)
    mensaje = strcat(PInterior, " (Si/No): ");
    disp(mensaje);
    resp = input("", "s");
  else
    mensaje = strcat(Preguntas(1,posCaract(1)), caracteristica(1), "? (Si/No): ");
    disp(mensaje);
    resp = input("", "s");
  end
  
  if(resp == 'Si')
    [x, y] = find(strcmp(candidatos, caracteristica));
    filas = x;
    eliminar = (setdiff([1:size(candidatos,1)],filas))';
  endif

  if(resp == 'No')
    [x,y] = find(strcmp(candidatos, caracteristica));
    eliminar = x;
  endif
  
  [estado, candidatos] = calcularEstado(candidatos, Opciones, eliminar);
  estado(posCaract)=0;
  
endwhile
msj = strcat("Estas pensando en: ", candidatos(1,1));
disp(msj);

File: Prototipo_Akinator.py
# -*- coding: utf-8 -*-
import numpy as np
import calcular_estado
import elegir_caracteristica


p_Pelo = "¿Su color de Pelo es "
p_Deporte = "¿Su Deporte favorito es "
p_Sexo = "¿Su sexo es "
p_Interior = "¿Vive en el Interior? "

opciones = np.array([["Rubio", "Marron", "Morocho"],["Futbol", "Rugby", "Tennis"],["Masculino", "Femenino", "Invalid"],["Si", "No", "Invalid"]])

preguntas = np.array([p_Pelo, p_Deporte, p_Sexo, p_Interior])

tabla = np.array([['Nombre', 'Pelo', 'Deporte', 'Sexo', 'Interior'],['Juan', opciones[0,0], opciones[1,0], opciones[2,0], opciones[3,1]],['Pablo', opciones[0,1], opciones[1,1], opciones[2,0], opciones[3,0]],['Maria', opciones[0,2], opciones[1,2], opciones[2,1], opciones[3,1]],['Guillermina', opciones[0,0], opciones[1,2], opciones[2,1], opciones[3,0]],['Ricardo', opciones[0,1], opciones[1,0], opciones[2,0], opciones[3,0]]])

repeticiones =  np.zeros((np.size(opciones,0),np.size(opciones,1)), dtype=int)
for i in range (0,np.size(opciones,0)):
    for j in range (0,np.size(opciones,1)):
        a = (opciones[i,j]==tabla)
        repeticiones[i,j] = sum(sum(a))

estado = repeticiones
candidatos = tabla[1:np.size(tabla,0),0:np.size(tabla,1)]

while np.size(candidatos,0) > 1:
    n = np.size(candidatos,0)
    print(candidatos)
    pos_caract = elegir_caracteristica.elegir_caracteristica(estado,n-1)
    caracterisitica = opciones[pos_caract[0],pos_caract[1]]

    if pos_caract[1] == 4:
        mensaje = p_Interior + ' (Si/No): '
        print(mensaje)
        resp = input("")
    else:
        mensaje = preguntas[pos_caract[0]] +' '+ caracterisitica + "? (Si/No): "
        print(mensaje)
        resp = input("")
    if resp == 'Si':
        indexes = np.argwhere(candidatos==caracterisitica)
        filas = np.zeros(len(indexes), dtype=int)
        contador = 0
        for j in indexes:
            filas[contador] = j[0]
            contador = contador + 1
        contador = 0
        ##ESTAMOS BIEN HASTA ACA
        eliminar = np.setdiff1d(np.arange(0,np.size(candidatos,0),1),filas)
       # eliminar = np.subtract(np.arange(0,np.size(candidatos,0),1),filas)
    if resp == 'No':
        indexes = np.argwhere(candidatos==caracterisitica)
        filas = np.zeros(len(indexes), dtype=int)
        contador = 0
        for j in indexes:
            filas[contador] = j[0]
            contador = contador + 1
        contador = 0
        eliminar = filas
    [estado, candidatos] = calcular_estado.calcular_estado(candidatos, opciones, eliminar)
    estado[pos_caract] = 0 

print("Estas pensando en: " + candidatos[0,0])

